flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSink.java

```diff
@@ -0,0 +1,124 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.elasticsearch.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.connector.sink.Committer;
+import org.apache.flink.api.connector.sink.GlobalCommitter;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.api.connector.sink.SinkWriter;
+import org.apache.flink.connector.base.DeliveryGuarantee;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+
+import org.apache.http.HttpHost;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Flink Sink to produce or update data in an Elasticsearch index. The sink supports the following
+ * delivery guarantees.
+ *
+ * <ul>
+ *   <li>{@link DeliveryGuarantee#NONE} does not provide any guarantees: actions are flushed to
+ *       Elasticsearch only depending on the configurations of the bulk processor. In case of a
+ *       failure it might happen that actions are lost if the bulk processor still has buffered
+ *       actions.
+ *   <li>{@link DeliveryGuarantee#AT_LEAST_ONCE} on a checkpoint the sink will wait until all
+ *       buffered actions are flushed to and acknowledged by Elasticsearch. No actions will be lost
+ *       but actions might be sent to Elasticsearch multiple times when Flink restarts.
+ * </ul>
+ *
+ * @param <IN> type of the records converted to Elasticsearch actions
+ * @see ElasticsearchSinkBuilder on how to construct a ElasticsearchSink
+ */
+@PublicEvolving
+public class ElasticsearchSink<IN> implements Sink<IN, Void, Void, Void> {
+
+    private final List<HttpHost> hosts;
+    private final ElasticsearchProcessor<IN> processor;
+    private final BulkProcessorConfig buildBulkProcessorConfig;
+    private final NetworkClientConfig networkClientConfig;
+    private final DeliveryGuarantee deliveryGuarantee;
+
+    ElasticsearchSink(
+            List<HttpHost> hosts,
+            ElasticsearchProcessor<IN> processor,
+            DeliveryGuarantee deliveryGuarantee,
+            BulkProcessorConfig buildBulkProcessorConfig,
+            NetworkClientConfig networkClientConfig) {
+        this.hosts = checkNotNull(hosts);
+        checkState(!hosts.isEmpty(), "Hosts cannot be empty.");
```

Comment:
`checkArg`

Reply:
I am still surprised when to use which precondition ... :)